pkg/segment/reader/segread/segreader.go

```diff
@@ -171,22 +171,32 @@ func (sfr *SegmentFileReader) loadBlockUsingBuffer(blockNum uint16) (bool, error
 	}
 }
 
+func (mcsr *MultiColSegmentReader) ValidateAndReadBlock(colsIndexMap map[int]struct{}, blockNum uint16) error {
+	for keyIndex := range colsIndexMap {
+		if keyIndex >= len(mcsr.allFileReaders) {
+			return fmt.Errorf("MultiColSegmentReader.ValidateAndReadBlock: keyIndex %v is out of bounds, len of allFileReaders: %v", keyIndex, len(mcsr.allFileReaders))
+		}
+
+		sfr := mcsr.allFileReaders[keyIndex]
+		if !sfr.isBlockLoaded || sfr.currBlockNum != blockNum {
+			valid, err := sfr.readBlock(blockNum)
+			if !valid {
+				return err
+			}
+			if err != nil {
+				return fmt.Errorf("MultiColSegmentReader.ValidateAndReadBlock: error loading blockNum: %v. Error: %+v", blockNum, err)
+			}
+		}
+	}
+
+	return nil
+}
+
 // returns the raw bytes of the blockNum:recordNum combination in the current segfile
 // optimized for subsequent calls to have the same blockNum
 // returns : encodedVal, error
 func (sfr *SegmentFileReader) ReadRecordFromBlock(blockNum uint16, recordNum uint16) ([]byte, error) {
 
-	if !sfr.isBlockLoaded || sfr.currBlockNum != blockNum {
-		valid, err := sfr.readBlock(blockNum)
-		if !valid {
-			return nil, err
-		}
-		if err != nil {
-			log.Errorf("SegmentFileReader.ReadRecordFromBlock: error loading blockNum: %v. Error: %+v", blockNum, err)
-			return nil, err
-		}
-	}
```

Comment:
Yeah I agree, I think this was the only check that would have optimized the code. Since, we already have a check to verify blockload and if blockNum and currBlockNum are equal, this function would mostly likely never call readBlock from ReadRecordFromBlock multiple times. If we keep the check anyways, I don't think there would be any performance benefit at all.

Reply:
I'm not sure I understand. You're saying you got a performance increase from removing this check? What impact did this have on performance?

Reply:
this check was executed for all the records taking around 1.5-2s for 100M records. Removing this check and validating at the block level was the main goal here.